src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -1756,32 +1756,50 @@ public static decimal Sum<TSource>(this IEnumerable<TSource> source, Func<TSourc
         public static int Min(this IEnumerable<int> source)
         {
             if (source == null) throw Error.ArgumentNull("source");
-            int value = 0;
-            bool hasValue = false;
-            foreach (int x in source)
+            int value;
+            using (IEnumerator<int> e = source.GetEnumerator())
             {
-                if (hasValue)
+                if (!e.MoveNext()) throw Error.NoElements();
+                value = e.Current;
+                while (e.MoveNext())
                 {
+                    int x = e.Current;
                     if (x < value) value = x;
                 }
-                else
-                {
-                    value = x;
-                    hasValue = true;
-                }
             }
-            if (hasValue) return value;
-            throw Error.NoElements();
+            return value;
         }
 
         public static int? Min(this IEnumerable<int?> source)
         {
             if (source == null) throw Error.ArgumentNull("source");
-            int? value = null;
-            foreach (int? x in source)
+            int? value;
+            using (IEnumerator<int?> e = source.GetEnumerator())
             {
-                if (value == null || x < value)
-                    value = x;
+                if (!e.MoveNext()) return null;
+                value = e.Current;
```

Comment:
Interestingly, the two cases where the effect of switching to `do…while` was most appreciable was the `int?` and `long?` versions of `Max` where it's actually a few percentage points in some cases. That difference drops a lot if the fast-path they have is taken out, so there seems to be some branching interaction there. Still, even the more expensive `do…while` is still a win on the old version, and consistency is to be favoured, so I'm putting it in there too.